File: KnapsackProblem/KnapsackProblem.py
import sys


def bf_recursion(bricks, capacity):
    if len(bricks) == 0 or capacity == 0:
        return 0
    if bricks[-1][1] > capacity:
        return bf_recursion(bricks[:-1], capacity)
    return max(bf_recursion(bricks[:-1], capacity), bf_recursion(bricks[:-1], capacity - bricks[-1][1]) + bricks[-1][0])


def brute_force(file_path):
    file_contents = open(file_path).read()[1:-3].split(', {')
    capacity = int(file_contents[0])
    bricks = [list(map(int, b.replace('{', '').split(', ')[:-1])) for b in file_contents[1:]]
    return bf_recursion(bricks, capacity)


def dp_recursion(bricks, capacity, cache, bricks_size):
    if bricks_size == 0 or capacity == 0:
        return 0
    if cache[capacity][bricks_size] != 0:
        return cache[capacity][bricks_size]
    if bricks[bricks_size - 1][1] > capacity:
        cache[capacity][bricks_size] = dp_recursion(bricks, capacity, cache, bricks_size - 1)
        return cache[capacity][bricks_size]
    first = dp_recursion(bricks, capacity, cache, bricks_size - 1)
    second = dp_recursion(
        bricks, capacity - bricks[bricks_size - 1][1], cache, bricks_size - 1
    ) + bricks[bricks_size - 1][0]
    if first > second:
        cache[capacity][bricks_size] = first
    else:
        cache[capacity][bricks_size] = second
    return cache[capacity][bricks_size]


def dynmaic_programming_recursion(file_path):
    # This solution works in solving the problem, however due to python's built in recursion limit, which we
    # cannot change, this will fail the 10th file and crash, it gets 0-9 correct though slightly slower.
    file_contents = open(file_path).read()[1:-3].split(', {')
    capacity = int(file_contents[0])
    bricks = [list(map(int, b.replace('{', '').split(', ')[:-1])) for b in file_contents[1:]]
    cache = [[0 for _ in range(len(bricks) + 1)] for _ in range(capacity + 1)]
    sys.setrecursionlimit(len(bricks) + 5)
    return dp_recursion(bricks, capacity, cache, len(bricks))


def dynamic_programming(file_path):
    file_contents = open(file_path).read()[1:-3].split(', {')
    capacity = int(file_contents[0])
    bricks = [list(map(int, b.replace('{', '').split(', ')[:-1])) for b in file_contents[1:]]
    cache = [[0 for _ in range(len(bricks) + 1)] for _ in range(capacity + 1)]
    for b_i in range(len(bricks) + 1):
        for c in range(capacity + 1):
            if b_i == 0 or c == 0:
                cache[c][b_i] = 0
            elif bricks[b_i - 1][1] > c:
                cache[c][b_i] = cache[c][b_i - 1]
            else:
                first = cache[c][b_i - 1]
                second = cache[c - bricks[b_i - 1][1]][b_i - 1] + bricks[b_i - 1][0]
                if first > second:
                    cache[c][b_i] = first
                else:
                    cache[c][b_i] = second
    return cache[capacity][len(bricks)]

File: KnapsackProblem/KnapsackProblem.cpp
#include <vector>
#include <string>
#include <fstream>
#include "KnapsackProblem.h"

using namespace std;

struct KnapsackProblem::Brick {
    int value;
    int weight;
};

vector<KnapsackProblem::Brick> KnapsackProblem::parseInput(const string &filePath) {
    ifstream file(filePath);
    string input;
    vector<Brick> bricks;
    Brick capacity{};
    if (file.is_open()) {
        while (getline(file, input, '{')) {
            if (!input.empty()) {
                if (input.size() > 7) {
                    bricks.push_back(
                        {stoi(input.substr(0, input.size() - 9)), stoi(input.substr(3, input.size() - 7))}
                    );
                } else {
                    capacity.value = stoi(input);
                }
            }
        }
        file.close();
    }
    bricks.push_back(capacity);
    return bricks;
}

int KnapsackProblem::bfRecursion(Brick* bricks, int bricksSize, int capacity) {
    if (!bricksSize || !capacity) {
        return 0;
    }
    if (bricks[bricksSize - 1].weight > capacity) {
        return bfRecursion(bricks, bricksSize - 1, capacity);
    }
    return max(
        bfRecursion(bricks, bricksSize - 1, capacity),
        bfRecursion(bricks, bricksSize - 1, capacity - bricks[bricksSize - 1].weight) + bricks[bricksSize - 1].value
    );
}

int KnapsackProblem::bruteForce(const string &filePath) {
    vector<Brick> bricksInput = parseInput(filePath);
    int capacity = bricksInput.back().value;
    bricksInput.pop_back();
    int bricksSize = int(bricksInput.size());
    Brick bricks[bricksSize];
    for (int i = 0; i < bricksSize; i++) {
        bricks[i] = bricksInput[i];
    }
    return bfRecursion(bricks, bricksSize, capacity);
}

int KnapsackProblem::dpRecursion(Brick* bricks, int bricksSize, int capacity, int** cache) {
    if (!bricksSize || !capacity) {
        return 0;
    }
    if (cache[capacity][bricksSize] != 0) {
        return cache[capacity][bricksSize];
    }
    if (bricks[bricksSize - 1].weight > capacity) {
        cache[capacity][bricksSize] = dpRecursion(bricks, bricksSize - 1, capacity, cache);
        return cache[capacity][bricksSize];
    }
    cache[capacity][bricksSize] = max(
        dpRecursion(bricks, bricksSize - 1, capacity, cache),
        dpRecursion(bricks, bricksSize - 1, capacity - bricks[bricksSize - 1].weight, cache) + bricks[bricksSize - 1].value
    );
    return cache[capacity][bricksSize];
}

int KnapsackProblem::dynamicProgrammingRecursion(const string &filePath) {
    vector<Brick> bricksInput = parseInput(filePath);
    int capacity = bricksInput.back().value;
    bricksInput.pop_back();
    int bricksSize = int(bricksInput.size());
    Brick bricks[bricksSize];
    for (int i = 0; i < bricksSize; i++) {
        bricks[i] = bricksInput[i];
    }
    int** cache = new int*[capacity + 1];
    for (int i = 0; i < capacity + 1; i++) {
        cache[i] = new int[bricksSize + 1];
        for (int j = 0; j < bricksSize + 1; j++) {
            cache[i][j] = 0;
        }
    }
    int ans = dpRecursion(bricks, bricksSize, capacity, cache);
    for (int i = 0; i <= capacity; i++) {
        delete [] cache[i];
    }
    delete [] cache;
    return ans;
}

int KnapsackProblem::dynamicProgramming(const string &filePath) {
    vector<Brick> bricksInput = parseInput(filePath);
    int capacity = bricksInput.back().value;
    bricksInput.pop_back();
    int bricksSize = int(bricksInput.size());
    Brick bricks[bricksSize];
    for (int i = 0; i < bricksSize; i++) {
        bricks[i] = bricksInput[i];
    }
    int** cache = new int*[capacity + 1];
    for (int i = 0; i < capacity + 1; i++) {
        cache[i] = new int[bricksSize + 1];
    }
    for (int x = 0; x < capacity + 1; x++) {
        for (int y = 0; y < bricksSize + 1; y++) {
            if (!x || !y) {
                cache[x][y] = 0;
            } else if (bricks[y - 1].weight > x) {
                cache[x][y] = cache[x][y - 1];
            } else {
                cache[x][y] = max(cache[x][y - 1], cache[x - bricks[y - 1].weight][y - 1] + bricks[y - 1].value);
            }
        }
    }
    int ans = cache[capacity][bricksSize];
    for (int i = 0; i < capacity + 1; i++) {
        delete [] cache[i];
    }
    delete [] cache;
    return ans;
}
